### PR TITLE
Auto show tabbar

### PR DESCRIPTION
**Pull request by artur-shaik**
Issue #771
Implemted common option for tab bar show strategy.
Options: always, never, multiple, switching.

### CODE REVIEW

File: qutebrowser/config/configdata.py [hunk 1]
```diff
@@ -471,13 +471,13 @@ def data(readonly=False):
              SettingValue(typ.LastClose(), 'ignore'),
              "Behavior when the last tab is closed."),
 
-            ('hide-auto',
-             SettingValue(typ.Bool(), 'false'),
-             "Hide the tab bar if only one tab is open."),
+            ('show',
+             SettingValue(typ.TabBarShow(), 'always'),
+             "The tab bar showing strategy."),
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: Something like "When to show the tab bar" would probably be more clear

File: qutebrowser/config/configdata.py [hunk 2]
```diff
@@ -471,13 +471,13 @@ def data(readonly=False):
...
 
-            ('hide-always',
-             SettingValue(typ.Bool(), 'false'),
-             "Always hide the tab bar."),
+            ('show-switching-delay',
+             SettingValue(typ.Int(), '800'),
+             "Time to show tab bar before hide it when tabs->show is switching."),
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: I'd prefer something like "Time to show the tab bar before hiding it when tabs->show is set to 'switching'." or so

File: qutebrowser/config/configtypes.py [hunk 3]
```diff
@@ -1547,3 +1547,14 @@ def complete(self):
              "curl 7.40.0")
         ]
         return out
+
+class TabBarShow(BaseType):
+
+    """How to format the question when downloading."""
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: This should be adjusted :wink:

File: qutebrowser/config/configtypes.py [hunk 4]
```diff
@@ -1547,3 +1547,14 @@ def complete(self):
...
+
+    valid_values = ValidValues(('always', "Always show the tab bar."),
+                               ('never', "Always hide the tab bar."),
+                               ('multiple', "Hide the tab bar if only one tab is open."),
+                               ('switching', "Show the tab bar only when current tab changed."))
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: "Show the tab bar when switching tabs." might be simpler

File: qutebrowser/mainwindow/tabwidget.py [hunk 5]
```diff
@@ -222,32 +223,48 @@ def __init__(self, win_id, parent=None):
         config_obj = objreg.get('config')
         config_obj.changed.connect(self.set_font)
         self.vertical = False
+        self.autoHideTimer = None
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabwidget.py]: Please call this `auto_hide_timer` instead - camel case is only used for Qt things. Also, it probably can be `_auto_hide_timer` (i.e. "private").

Can you please also add a small description to the docstring under `Attributes:`?

File: qutebrowser/mainwindow/tabwidget.py [hunk 6]
```diff
@@ -222,32 +223,48 @@ def __init__(self, win_id, parent=None):
...
 
     def _tabhide(self):
         """Hide the tab bar if needed."""
-        hide_auto = config.get('tabs', 'hide-auto')
-        hide_always = config.get('tabs', 'hide-always')
-        if hide_always or (hide_auto and self.count() == 1):
+        show = config.get('tabs', 'show')
+        showNever = show == 'never'
+        showSwitching = show == 'switching'
+        showMultiple = show == 'multiple'
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabwidget.py]: Please also call these `show_never/switching/multiple`.

**Comment by The-Compiler** [comment]
I think the way you're setting the timer to `None` in between is rather confusing - it's probably easier to just set it to `QTimer()` and use `.setSingleShot(True)` once in `__init__`, and then make `on_change` just call `setInterval()` and `start()` (and `stop()` instead of setting it to `None`).

There also should be some kind of migration for existing configs, but I'll look into that myself as it's not really set up to merge two options into one so far.

**Comment by artur-shaik** [comment]
Ok, thanks for comments. Now again I see my weird code in some places :smile:. Will change it.

**Comment by artur-shaik** [comment]
I hope nothing is forgotten.

**Comment by The-Compiler** [comment]
Sorry for the delay - I'll merge this after `v0.3` is out as I don't want to delay v0.3 any further.

I'm sorry I have to do this to you again, like with the visual mode and v0.2 already :wink: - but hopefully I'll release more often in the future.

**Comment by artur-shaik** [comment]
No matter, I already use it :smile:, thanks.

**Comment by The-Compiler** [comment]
I decided to simply mark the old options as deleted - as I'll rewrite much of the config system anyways it doesn't make sense to add some sophisticated migration now. Users just will have to re-set the value according to their taste :wink:

I also cherry-picked the two relevant commits instead of merging the PR to avoid all the unnecessary merge commits - so I'll close this as the code is in.

Thank you! :smile:

**Pull request closed without merging**